src/channels.rs

```diff
@@ -203,6 +279,16 @@ impl Year {
         self.courses.values().find_map(|x| x.courses.get(role_name))
     }
 
+    fn get_semester_name(&self, role_name: &str) -> Option<String> {
```

Comment:
if the signature is `<'s>(&'s self, role_name: &str) -> Option<&'s str>` does the code compile without the `clone()`? If yes then you can avoid allocating here

Reply:
I had that in the beginning, the problem is that in the `rename_course` function I would have an error because of a mutable after an immutable one, so I decided to clone.
```rs
error[E0502]: cannot borrow `*self` as mutable because it is also borrowed as immutable
   --> src/channels.rs:211:13
    |
210 |         if let Some((year, semester)) = self.get_year_semester_names(course) {
    |                                         ---- immutable borrow occurs here
211 |             self.move_course(course, &year, &semester, Some(new_name), ctx, guild)
    |             ^^^^^-----------^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |             |    |
    |             |    immutable borrow later used by call
    |             mutable borrow occurs here
```

Reply:
ah frick, that's right. It could create a dangling ptr if not cloned